qiskit/aqua/operators/list_ops/summed_op.py

```diff
@@ -142,3 +150,43 @@ def to_legacy_op(self, massive: bool = False) -> LegacyBaseOperator:
             coeff = cast(float, self.coeff)
 
         return self.combo_fn(legacy_ops) * coeff
+
+    def equals(self, other: OperatorBase) -> bool:
+        """Check if other is equal to self.
+
+        Args:
+            other: The other operator to check for equality.
+
+        Returns:
+            True, if other and self are equal, otherwise False.
+
+        Examples:
+            >>> from qiskit.aqua.operators import X, Z
+            >>> 2 * X == X + X
+            True
+            >>> X + Z == Z + X
+            True
+        """
+        self_reduced, other_reduced = self.reduce(), other.reduce()
+        if not isinstance(other_reduced, type(self_reduced)):
+            return False
+
+        # check if reduced op is still a SummedOp
+        if not isinstance(self_reduced, SummedOp):
+            return self_reduced == other_reduced
+
+        self_reduced = cast(SummedOp, self_reduced)
+        other_reduced = cast(SummedOp, other_reduced)
+        if not len(self_reduced.oplist) == len(other_reduced.oplist):
+            return False
+
+        # absorb coeffs into the operators
+        if self_reduced.coeff != 1:
```

Comment:
If both coeffs were not a value of 1 would it not be a bit more efficient just to scale one set opertaors coeffs by an appropriate factor to match the other rather than do a similar thing to both? I do not know at what level of decimal places for each coeff we would want to consider these equal. The set relies on hash of course - but is that not just in the pauli_op and some string repr? Since a summed_op can potentially contain other primitive_ops, not just paulis, is the equality based on set here bogus?

Reply:
Just double checked, the set comparison must be fixed!
```python
from qiskit.aqua.operators import Z, MatrixOp
z = MatrixOp([[1, 0], [0, -1]])
z + Z == Z + z  # TypeError: unhashable type: 'MatrixOp'
```
My first thought would be to add a hash method to all the PrimitiveOps, what do you think?

Reply:
While it may be fine to add hash to all primitive ops, given they are logically immutable, the set thing still does not seem like it will provide a complete solution when the summed op can contain a mix of primitives. A set of MatrixOps may sum to the same even when their set of individual matrices are different - no?
And I think that the objects ideally need their own defined equals  - otherwise for more complex stuff I guess CircuitOps would most likely be equal only when they refer to the exact same object. Maybe its a limitation that is acceptable in this context.

Reply:
Yes, it's will not cover all possible cases, though by calling `reduce` hopefully some of these can be covered.